### PR TITLE
Feature: load metadata

### PR DESCRIPTION
**Pull request by Zorin95670**

### CODE REVIEW

**Comment by k-y0u** [comment]
the github icon is weird on github...

File: src/models/GenericComponentDefinition.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { ComponentDefinition } from 'leto-modelizer-plugin-core';
+
+/**
+ * Generic component definition.
+ */
+class GenericComponentDefinition extends ComponentDefinition {
+  /**
+   * Default constructor.
+   * @param {Object} [props] - Object that contains all properties to set.
+   * @param {boolean} [props.available=false] - Is these definitions is available to be used by end
```
Comment (justinehell) [inline src/models/GenericComponentDefinition.js:10]: '**If** these definitions **are**...' or '**Are** these definitions available...'

File: src/models/GenericComponentDefinition.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+   * user.
+   */
+  constructor(props = { available: false }) {
+    super(props);
+    /**
+     * Is these definitions is available to be used by end user.
```
Comment (justinehell) [inline src/models/GenericComponentDefinition.js:16]: same here

File: src/models/TriggerDefinition.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import { ComponentDefinition } from 'leto-modelizer-plugin-core';
+
+/**
+ * Trigger definition.
+ */
+class TriggerDefinition extends ComponentDefinition {
+  /**
+   * Default constructor.
+   * @param {Object} [props] - Object that contains all properties to set.
+   * @param {String} [props.action] - Action name call by the trigger.
```
Comment (justinehell) [inline src/models/TriggerDefinition.js:10]: Suggestion : `Action name called if triggered`

File: src/models/TriggerDefinition.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+   */
+  constructor(props = {
+    action: null,
+  }) {
+    super(props);
+    /**
+     * Action name call by the trigger.
```
Comment (justinehell) [inline src/models/TriggerDefinition.js:17]: same

File: src/models/GenericComponentDefinition.js (hunk 1 above)
Comment (k-y0u) [inline src/models/GenericComponentDefinition.js:10]: is...is... ? to be used by end ?
What do you mean ?

Comment (k-y0u) [reply src/models/GenericComponentDefinition.js:10]: dot missing

File: src/models/GenericComponentDefinition.js (hunk 2 above)
Comment (k-y0u) [inline src/models/GenericComponentDefinition.js:16]: `Are these definitions can be used by (the) end user(s).` (choose only one of the parentheses)

File: src/models/TriggerDefinition.js (hunk 3 above)
Comment (k-y0u) [inline src/models/TriggerDefinition.js:10]: called

File: src/models/GenericComponentDefinition.js (hunk 1 above)
Comment (k-y0u) [inline src/models/GenericComponentDefinition.js:10]: boolean or Boolean ?

File: src/models/TriggerDefinition.js (hunk 3 above)
Comment (k-y0u) [inline src/models/TriggerDefinition.js:10]: String or string ?

File: src/models/TriggerDefinition.js (hunk 4 above)
Comment (k-y0u) [inline src/models/TriggerDefinition.js:17]: called

File: tests/unit/metadata/GithubActionMetadata.spec.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import GithubActionMetadata from 'src/metadata/GithubActionMetadata';
+import { DefaultData } from 'leto-modelizer-plugin-core';
+import fs from 'fs';
+
+describe('Test GithubActionMetadata', () => {
+  describe('Test methods', () => {
+    describe('Test method: parse', () => {
+      it('Default metadata should return valid content', () => {
```
Comment (justinehell) [inline tests/unit/metadata/GithubActionMetadata.spec.js:8]: suggestion `Should return valid content when parsing default metadata`

File: changelog.md [hunk 6]
```diff
@@ -9,5 +9,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
-- Setup GitHub workflow
-- Setup default plugin structure
+- Setup GitHub workflow.
+- Setup default plugin structure.
+- Setup first version of metadata
```
Comment (justinehell) [inline changelog.md:14]: `.` or no `.` ?

File: tests/unit/models/GenericComponentDefinition.spec.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+describe('Test class: GenericComponentDefinition', () => {
+  describe('Test constructor', () => {
+    it('Check variable instantiation', () => {
+      const component = new GenericComponentDefinition();
+
+      expect(component.available).toEqual(false);
+    });
+
+    it('Check passing undefined variables to constructor', () => {
```
Comment (k-y0u) [inline tests/unit/models/GenericComponentDefinition.spec.js:11]: not an `undefined variables`, it's an empty object (so it's defined).

**Review by justinehell — CHANGES_REQUESTED** [review]

File: tests/unit/models/GenericComponentDefinition.spec.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+      const component = new GenericComponentDefinition({});
+
+      expect(component.available).toEqual(false);
+    });
+
+    it('Check passing all variables to constructor', () => {
```
Comment (k-y0u) [inline tests/unit/models/GenericComponentDefinition.spec.js:17]: `all variables`... only one, lol

File: tests/unit/models/TriggerDefinition.spec.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+describe('Test class: TriggerDefinition', () => {
+  describe('Test constructor', () => {
+    it('Check variable instantiation', () => {
+      const trigger = new TriggerDefinition();
+
+      expect(trigger.action).toBeNull();
+    });
+
+    it('Check passing undefined variables to constructor', () => {
```
Comment (k-y0u) [inline tests/unit/models/TriggerDefinition.spec.js:11]: same

File: tests/unit/models/TriggerDefinition.spec.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+      const trigger = new TriggerDefinition({});
+
+      expect(trigger.action).toBeNull();
+    });
+
+    it('Check passing all variables to constructor', () => {
```
Comment (k-y0u) [inline tests/unit/models/TriggerDefinition.spec.js:17]: same

File: tests/unit/models/GenericComponentDefinition.spec.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
+import GenericComponentDefinition from 'src/models/GenericComponentDefinition';
+
+describe('Test class: GenericComponentDefinition', () => {
+  describe('Test constructor', () => {
+    it('Check variable instantiation', () => {
```
Comment (k-y0u) [inline tests/unit/models/GenericComponentDefinition.spec.js:5]: which variable has been instantiated here ? without any parameter given....

File: tests/unit/models/TriggerDefinition.spec.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
+import TriggerDefinition from 'src/models/TriggerDefinition';
+
+describe('Test class: TriggerDefinition', () => {
+  describe('Test constructor', () => {
+    it('Check variable instantiation', () => {
```
Comment (k-y0u) [inline tests/unit/models/TriggerDefinition.spec.js:5]: same

File: src/metadata/GithubActionMetadata.js [hunk 13]
```diff
@@ -1,21 +1,92 @@
-import { DefaultMetadata } from 'leto-modelizer-plugin-core';
+import {
+  ComponentAttributeDefinition, ComponentDefinition,
```
Comment (k-y0u) [inline src/metadata/GithubActionMetadata.js:2]: divide

File: src/metadata/GithubActionMetadata.js [hunk 14]
```diff
@@ -1,21 +1,92 @@
...
+  DefaultMetadata,
+} from 'leto-modelizer-plugin-core';
+import metadata from 'src/assets/metadata/index';
+import GenericComponentDefinition from 'src/models/GenericComponentDefinition';
+import TriggerDefinition from 'src/models/TriggerDefinition';
 
 /**
  * Class to validate and retrieve components definitions from GithubAction metadata.
```
Comment (k-y0u) [inline src/metadata/GithubActionMetadata.js:10]: is there multiple definition for one components ?

File: src/metadata/GithubActionMetadata.js [hunk 15]
```diff
@@ -1,21 +1,92 @@
...
  */
 class GithubActionMetadata extends DefaultMetadata {
-  validate() {
-    return super.validate();
-  }
-
   /**
    * Parse all components definitions from metadata.
```
Comment (k-y0u) [inline src/metadata/GithubActionMetadata.js:14]: same

File: src/assets/metadata/generic.json [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+    }]
+  },
+  "job": {
+    "type": "job",
+    "model": "DefaultContainer",
+    "icon": "DefaultJob",
+    "available": true,
+    "parentTypes": ["workflow"],
+    "childrenTypes": ["step"],
+    "attributes": [{
```
Comment (Zorin95670) [inline src/assets/metadata/generic.json:32]: Missing attributes `name` and `runs-on`

File: src/metadata/GithubActionMetadata.js [hunk 17]
```diff
@@ -1,21 +1,92 @@
...
   parse() {
+    const generics = this.parseGeneric();
+    const others = [];
+
+    Object.keys(metadata)
+      .filter((key) => key !== 'generic')
+      .flatMap((key) => metadata[key])
+      .forEach((data) => {
+        let definition = null;
+        if (data.type === 'trigger') {
```
Comment (k-y0u) [inline src/metadata/GithubActionMetadata.js:25]: jump

File: src/metadata/GithubActionMetadata.js [hunk 18]
```diff
@@ -1,21 +1,92 @@
...
+          definition = new TriggerDefinition({
+            ...generics.find((def) => def.type === data.type),
+          });
+        } else {
+          definition = new ComponentDefinition({
+            ...generics.find((def) => def.type === data.type),
+          });
+        }
+        Object.keys(data)
```
Comment (k-y0u) [inline src/metadata/GithubActionMetadata.js:34]: jump

File: src/metadata/GithubActionMetadata.js [hunk 19]
```diff
@@ -1,21 +1,92 @@
...
-  validate() {
-    return super.validate();
-  }
-
   /**
    * Parse all components definitions from metadata.
    */
   parse() {
+    const generics = this.parseGeneric();
+    const others = [];
```
Comment (k-y0u) [inline src/metadata/GithubActionMetadata.js:18]: function parseOther()

**Review by k-y0u — DISMISSED** [review]

File: src/models/GenericComponentDefinition.js (hunk 1 above)
Comment (Zorin95670) [reply src/models/GenericComponentDefinition.js:10]: done

File: src/models/GenericComponentDefinition.js (hunk 2 above)
Comment (Zorin95670) [reply src/models/GenericComponentDefinition.js:16]: done

File: src/models/TriggerDefinition.js (hunk 3 above)
Comment (Zorin95670) [reply src/models/TriggerDefinition.js:10]: done

File: tests/unit/metadata/GithubActionMetadata.spec.js (hunk 5 above)
Comment (Zorin95670) [reply tests/unit/metadata/GithubActionMetadata.spec.js:8]: done

File: src/models/TriggerDefinition.js (hunk 4 above)
Comment (Zorin95670) [reply src/models/TriggerDefinition.js:17]: done

File: changelog.md (hunk 6 above)
Comment (Zorin95670) [reply changelog.md:14]: done

File: src/assets/metadata/generic.json (hunk 16 above)
Comment (Zorin95670) [reply src/assets/metadata/generic.json:32]: done

File: src/models/TriggerDefinition.js (hunk 3 above)
Comment (Zorin95670) [reply src/models/TriggerDefinition.js:10]: done

File: src/models/GenericComponentDefinition.js (hunk 1 above)
Comment (Zorin95670) [reply src/models/GenericComponentDefinition.js:10]: done

File: src/models/TriggerDefinition.js (hunk 3 above)
Comment (Zorin95670) [reply src/models/TriggerDefinition.js:10]: we are waiting for the pr that add the jsdoc ;)

File: src/models/TriggerDefinition.js (hunk 4 above)
Comment (Zorin95670) [reply src/models/TriggerDefinition.js:17]: done

File: tests/unit/models/GenericComponentDefinition.spec.js (hunk 7 above)
Comment (Zorin95670) [reply tests/unit/models/GenericComponentDefinition.spec.js:11]: done

File: tests/unit/models/GenericComponentDefinition.spec.js (hunk 8 above)
Comment (Zorin95670) [reply tests/unit/models/GenericComponentDefinition.spec.js:17]: keep it :)

File: tests/unit/models/TriggerDefinition.spec.js (hunk 10 above)
Comment (Zorin95670) [reply tests/unit/models/TriggerDefinition.spec.js:17]: keep it :)

File: tests/unit/models/GenericComponentDefinition.spec.js (hunk 11 above)
Comment (Zorin95670) [reply tests/unit/models/GenericComponentDefinition.spec.js:5]: done

File: src/metadata/GithubActionMetadata.js (hunk 13 above)
Comment (Zorin95670) [reply src/metadata/GithubActionMetadata.js:2]: done

File: src/metadata/GithubActionMetadata.js (hunk 14 above)
Comment (Zorin95670) [reply src/metadata/GithubActionMetadata.js:10]: yes

File: src/metadata/GithubActionMetadata.js (hunk 15 above)
Comment (Zorin95670) [reply src/metadata/GithubActionMetadata.js:14]: done

File: src/metadata/GithubActionMetadata.js (hunk 17 above)
Comment (Zorin95670) [reply src/metadata/GithubActionMetadata.js:25]: yes 
image

File: src/models/GenericComponentDefinition.js [hunk 20]
```diff
@@ -0,0 +1,23 @@
+import { ComponentDefinition } from 'leto-modelizer-plugin-core';
+
+/**
+ * Generic component definition.
+ */
+class GenericComponentDefinition extends ComponentDefinition {
+  /**
+   * Default constructor.
+   * @param {Object} [props] - Object that contains all properties to set.
+   * @param {Boolean} [props.available=false] - If these definition is available to be used by end
```
Comment (justinehell) [inline src/models/GenericComponentDefinition.js:10]: `If this definition is...` for singular

File: src/models/GenericComponentDefinition.js [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+   * user(s).
+   */
+  constructor(props = { available: false }) {
+    super(props);
+    /**
+     * If these definition is available to be used by end user(s).
```
Comment (justinehell) [inline src/models/GenericComponentDefinition.js:16]: same

**Review by justinehell — CHANGES_REQUESTED** [review]

File: src/models/GenericComponentDefinition.js (hunk 1 above)
Comment (justinehell) [reply src/models/GenericComponentDefinition.js]: `If this definition is...` for singular

File: src/models/GenericComponentDefinition.js (hunk 2 above)
Comment (justinehell) [reply src/models/GenericComponentDefinition.js]: change here too `these definitions` -> `this definition`

**Review by justinehell — APPROVED** [review]

**Comment by k-y0u** [comment]
> the github icon is weird on github...

Not same size as other icons.

File: src/models/GenericComponentDefinition.js (hunk 1 above)
Comment (k-y0u) [reply src/models/GenericComponentDefinition.js]: if the line is less than 100 characters, it's ok.

File: tests/unit/models/GenericComponentDefinition.spec.js (hunk 8 above)
Comment (k-y0u) [reply tests/unit/models/GenericComponentDefinition.spec.js:17]: passing props to constructor

Comment (Zorin95670) [reply tests/unit/models/GenericComponentDefinition.spec.js:17]: done

File: tests/unit/models/TriggerDefinition.spec.js (hunk 10 above)
Comment (Zorin95670) [reply tests/unit/models/TriggerDefinition.spec.js:17]: done

**Review by k-y0u — APPROVED** [review]

**Review by justinehell — APPROVED** [review]

**Pull request merged**